modules/core/src/main/java/org/apache/ignite/internal/processors/query/aware/IndexRebuildStateStorage.java

```diff
@@ -0,0 +1,298 @@
+ */
+
+package org.apache.ignite.internal.processors.query.aware;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.cache.GridCacheProcessor;
+import org.apache.ignite.internal.processors.cache.persistence.GridCacheDatabaseSharedManager;
+import org.apache.ignite.internal.processors.cache.persistence.IgniteCacheDatabaseSharedManager;
+import org.apache.ignite.internal.processors.cache.persistence.checkpoint.CheckpointListener;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.MetaStorage;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.MetastorageLifecycleListener;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.ReadOnlyMetastorage;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.ReadWriteMetastorage;
+import org.apache.ignite.internal.processors.query.GridQueryProcessor;
+import org.apache.ignite.internal.util.GridBusyLock;
+import org.apache.ignite.internal.util.lang.IgniteThrowableConsumer;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+
+import static org.apache.ignite.internal.processors.query.aware.IndexRebuildState.State.COMPLETED;
+import static org.apache.ignite.internal.processors.query.aware.IndexRebuildState.State.DELETE;
+import static org.apache.ignite.internal.processors.query.aware.IndexRebuildState.State.INIT;
+
+/**
+ * Holder of up-to-date information about rebuilding cache indexes.
+ * Helps to avoid the situation when the index rebuilding process is interrupted
+ * and after a node restart/reactivation the indexes will become inconsistent.
+ * <p/>
+ * To do this, before rebuilding the indexes, call {@link #onStartRebuildIndexes}
+ * and after it {@link #onFinishRebuildIndexes}. Use {@link #completed} to check
+ * if the index rebuild has completed.
+ * <p/>
+ * To prevent leaks, it is necessary to use {@link #onFinishRebuildIndexes}
+ * when detecting the fact of destroying the cache.
+ */
+public class IndexRebuildStateStorage implements MetastorageLifecycleListener, CheckpointListener {
+    /** Key prefix for the MetaStorage. */
+    public static final String KEY_PREFIX = "rebuild-sql-indexes-";
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** MetaStorage synchronization mutex. */
+    private final Object metaStorageMux = new Object();
+
+    /** Node stop lock. */
+    private final GridBusyLock stopNodeLock = new GridBusyLock();
+
+    /** Current states. Mapping: cache name -> index rebuild state. */
+    private final ConcurrentMap<String, IndexRebuildState> states = new ConcurrentHashMap<>();
```

Comment:
seems that _states_ are never cleans ?

Reply:
_states_ is cleared here:
- IndexRebuildStateStorage#onFinishRebuildIndexes - Using a ConcurrentMap#compute that returns **null** for non-persistent cache, thereby deletes the entry in the _states_;
- IndexRebuildStateStorage#afterCheckpointEnd - Uses the same ConcurrentMap#compute, but removes only those with status IndexRebuildState.State#DELETE.

I also added a check to the ResumeRebuildIndexTest#testDeleteIndexRebuildStateOnDestroyCache that the _states_ is cleared.